crepecake/src/main/java/com/google/cloud/tools/crepecake/http/Connection.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.crepecake.http;
+
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpMethods;
+import com.google.api.client.http.HttpRequestFactory;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.URL;
+
+/** Sends an HTTP {@link Request} and stores the {@link Response}. */
+public class Connection implements Closeable {
+
+  private static final HttpRequestFactory HTTP_REQUEST_FACTORY =
+      new NetHttpTransport().createRequestFactory();
+
+  private final HttpRequestFactory requestFactory;
+
+  private HttpResponse httpResponse;
+
+  /** The URL to send the request to. */
+  private GenericUrl url;
+
+  /**
+   * Make sure to wrap with a try-with-resource to ensure that the connection is closed after usage.
+   */
+  public Connection(URL url) throws IOException {
+    this(url, HTTP_REQUEST_FACTORY);
+  }
+
+  @VisibleForTesting
+  Connection(URL url, HttpRequestFactory requestFactory) {
+    this.url = new GenericUrl(url);
+    this.requestFactory = requestFactory;
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (httpResponse == null) {
+      return;
+    }
+
+    httpResponse.disconnect();
+  }
+
+  /** Sends the request with method GET. */
+  public Response get(Request request) throws IOException {
+    return send(HttpMethods.GET, request);
+  }
+
+  /** Sends the request with method POST. */
+  public Response post(Request request) throws IOException {
+    return send(HttpMethods.POST, request);
+  }
+
+  /** Sends the request with method PUT. */
+  public Response put(Request request) throws IOException {
+    return send(HttpMethods.PUT, request);
+  }
+
+  /** Sends the request. */
+  private Response send(String httpMethod, Request request) throws IOException {
+    httpResponse =
```

Comment:
Instead of implementing Closeable, could we simply close httpResponse here before returning?

Reply:
According to the doc of `disconnect`:

```
#disconnect: Close the HTTP response content using {@link #ignore}
#ignore: Closes the content of the HTTP response from {@link #getContent()}, ignoring any content.
```

When disconnect is called, the response is essentially considered dead and nothing can be retrieved from it. Therefore, we can't close the httpResponse before its content is actually read via the returned `Response` object.

So, the use of `Connection` would be like:

```
try (Connection connection = new Connection(url)) {
  Response response = connection.get(request);
  // ... process the response
}
```